File: apps/android/app/src/main/java/com/ailimits/dashboard/provider/ZaiProvider.kt
package com.ailimits.dashboard.provider

import com.ailimits.dashboard.data.ProviderConfig
import com.ailimits.dashboard.data.ProviderType
import com.ailimits.dashboard.data.UsageSnapshot
import com.ailimits.dashboard.data.UsageStatus

class ZaiProvider : UsageProvider {
    override fun validateConfig(config: ProviderConfig): Boolean {
        if (config.type != ProviderType.ZAI) return false
        if (config.customEndpoint != null) return true
        return config.manualUsed != null &&
            config.manualLimit != null &&
            config.manualLimit!! > 0
    }

    override suspend fun fetchUsage(config: ProviderConfig): UsageSnapshot {
        if (config.type != ProviderType.ZAI) return errorSnapshot(config.id)
        config.customEndpoint?.let { return fetchFromCustom(config) }
        config.manualUsed?.let { used ->
            config.manualLimit?.let { limit ->
                if (limit > 0) {
                    val ratio = used / limit
                    return UsageSnapshot(
                        providerId = config.id,
                        used = used,
                        limit = limit,
                        unit = if (config.displayName?.contains("%") == true) "%" else "manual",
                        timestamp = System.currentTimeMillis(),
                        status = usageStatus(ratio),
                        displayName = config.displayName ?: "z.ai",
                    )
                }
            }
        }
        return errorSnapshot(config.id)
    }

    private suspend fun fetchFromCustom(config: ProviderConfig): UsageSnapshot =
        kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
            try {
                val url = java.net.URL(config.customEndpoint!!)
                (url.openConnection() as javax.net.ssl.HttpsURLConnection).apply {
                    requestMethod = "GET"
                    config.apiKey?.let { setRequestProperty("Authorization", "Bearer $it") }
                    connectTimeout = 10000
                    readTimeout = 10000
                }.run {
                    if (responseCode != 200) return@withContext errorSnapshot(config.id)
                    val body = inputStream.bufferedReader().use { it.readText() }
                    val json = org.json.JSONObject(body)
                    val used = json.optDouble("used", 0.0)
                    val limit = json.optDouble("limit", 1.0).takeIf { it > 0 } ?: 1.0
                    UsageSnapshot(
                        providerId = config.id,
                        used = used,
                        limit = limit,
                        unit = json.optString("unit", "manual"),
                        timestamp = System.currentTimeMillis(),
                        status = usageStatus(used / limit),
                        displayName = config.displayName ?: "z.ai",
                    )
                }
            } catch (_: Exception) {
                errorSnapshot(config.id)
            }
        }

    private fun errorSnapshot(providerId: String) = UsageSnapshot(
        providerId = providerId,
        used = 0.0,
        limit = 1.0,
        unit = "manual",
        timestamp = System.currentTimeMillis(),
        status = UsageStatus.ERROR,
        displayName = "z.ai",
    )
}

File: packages/core/src/providers/ZaiProvider.ts
import type { ProviderConfig } from "../models/ProviderConfig.js";
import type { UsageSnapshot } from "../models/UsageSnapshot.js";
import type { IUsageProvider } from "./IUsageProvider.js";

interface CustomEndpointResponse {
  used?: number;
  limit?: number;
  unit?: string;
}

/**
 * z.ai — свободная форма: ручной ввод или кастомный endpoint.
 */
export class ZaiProvider implements IUsageProvider {
  validateConfig(config: ProviderConfig): boolean {
    if (config.type !== "zai") return false;
    if (config.customEndpoint) return true;
    const manual =
      config.manualUsed != null &&
      config.manualLimit != null &&
      config.manualLimit > 0;
    return manual;
  }

  async fetchUsage(config: ProviderConfig): Promise<UsageSnapshot> {
    if (config.type !== "zai") {
      return this.errorSnapshot(config.id);
    }
    if (config.customEndpoint) {
      return this.fetchFromCustom(config);
    }
    if (
      config.manualUsed != null &&
      config.manualLimit != null &&
      config.manualLimit > 0
    ) {
      const used = config.manualUsed;
      const limit = config.manualLimit;
      const ratio = used / limit;
      const status = ratio >= 1 ? "error" : ratio >= 0.8 ? "warning" : "ok";
      return {
        providerId: config.id,
        used,
        limit,
        unit: config.displayName?.includes("%") ? "%" : "manual",
        timestamp: Date.now(),
        status,
        displayName: config.displayName ?? "z.ai",
      };
    }
    return this.errorSnapshot(config.id);
  }

  private async fetchFromCustom(config: ProviderConfig): Promise<UsageSnapshot> {
    const url = config.customEndpoint!;
    try {
      const headers: Record<string, string> = {};
      if (config.apiKey) headers["Authorization"] = `Bearer ${config.apiKey}`;
      const res = await fetch(url, { headers });
      if (!res.ok) {
        return this.errorSnapshot(config.id);
      }
      const body = (await res.json()) as CustomEndpointResponse;
      const used = body.used ?? 0;
      const limit = body.limit ?? 1;
      const ratio = limit > 0 ? used / limit : 0;
      const status = ratio >= 1 ? "error" : ratio >= 0.8 ? "warning" : "ok";
      return {
        providerId: config.id,
        used,
        limit,
        unit: body.unit ?? "manual",
        timestamp: Date.now(),
        status,
        displayName: config.displayName ?? "z.ai",
      };
    } catch {
      return this.errorSnapshot(config.id);
    }
  }

  private errorSnapshot(providerId: string): UsageSnapshot {
    return {
      providerId,
      used: 0,
      limit: 1,
      unit: "manual",
      timestamp: Date.now(),
      status: "error",
      displayName: "z.ai",
    };
  }
}
